### PR TITLE
Update bloom filter

### PR DESCRIPTION
**Pull request by willclarktech**
POC translation of https://github.com/OpenMined/PyPSI/pull/3

It's probably better to implement the bitarray ourselves than rely on this library. ~~Also I just used native `Buffer` because we haven't finalised the distribution process yet.~~

### CODE REVIEW

File: src/dataStructure/bloomFilter/index.ts [hunk 1]
```diff
@@ -1,15 +1,105 @@
...
+  return Math.ceil(bitLength / 8) * 8
+}
+
+const getHash = (element: string, bitLength: number, nonce: number): number => {
+  const hashObject = sha256.create()
+  hashObject.update(
+    Buffer.concat([
+      Buffer.from([nonce]),
+      Buffer.from(element, 'latin1')
+    ]).toString('latin1')
```
Comment (s0l0ist) [inline src/dataStructure/bloomFilter/index.ts:43]: can we replace 'latin1' with it's more descriptive alias, 'binary'?

File: src/dataStructure/bloomFilter/index.ts [hunk 2]
```diff
@@ -1,15 +1,100 @@
...
+    capacity = MAX_ELEMENTS,
+    falsePositiveProbability = FALSE_POSITIVE_PROBABILITY
+  ) => {
+    let count = 0
+    const bitLength = getBitLength(capacity, falsePositiveProbability)
+    const bitArray = BitArray.factory(0, bitLength)
+    const numHashFunctions = Math.ceil(-Math.log2(falsePositiveProbability))
+
+    const getHashes = (element: string): readonly number[] => {
+      const [h1, h2] = NONCES.map(getHash.bind(null, element, bitLength))
```
Comment (s0l0ist) [inline src/dataStructure/bloomFilter/index.ts:57]: I'm a bit confused as to why we need to `bind` getHash here. Couldn't we just pass in a partial to `NONCES.map?`

Ex: 

	const getHash = (element: string, bitLength: number, nonce: number): number => {
	  const hashObject = sha256.create()
	  hashObject.update(
	    Buffer.concat([
	      Buffer.from([nonce]),
	      Buffer.from(element, 'binary')
	    ]).toString('binary')
	  )
	  const rawHash = hashObject.digest().toHex()
	  const bigIntHash = bigInt(rawHash, 16)
	  // Non-uniformity is negligible for Bloom filters of reasonable size
	  return bigIntHash.mod(bitLength).toJSNumber()
	}

    // Define a partially applied getHash helper
    const getHashPartial = (element, bitLength) => nonce => getHash(element, bitLength, nonce)

	const bloomFilter: BloomFilterDataStructure = {
	  create: (
	    BloomFilter.create(length, errorRate),	    capacity = MAX_ELEMENTS,
	  from: (items, errorRate = ERROR_RATE) => BloomFilter.from(items, errorRate)	    falsePositiveProbability = FALSE_POSITIVE_PROBABILITY
	  ) => {
	    let count = 0
	    const bitLength = getBitLength(capacity, falsePositiveProbability)
	    const bitArray = BitArray.factory(0, bitLength)
	    const numHashFunctions = Math.ceil(-Math.log2(falsePositiveProbability))

	    const getHashes = (element: string): readonly number[] => {
      	  const [h1, h2] = NONCES.map(getHashPartial(element, bitLength))
	}

Comment (willclarktech) [reply src/dataStructure/bloomFilter/index.ts:57]: I used `.bind` here to perform partial application, not to set the context. We don't need to define a new function for partial application.

Comment (s0l0ist) [reply src/dataStructure/bloomFilter/index.ts:57]: Ah yes, it's a nice little feature of `bind` that I don't often see.

**Pull request merged**